### PR TITLE
feat: Unit page - create header

### PR DESCRIPTION
**Pull request by ihor-romaniuk**
**Description:** unit page header

**Youtrack:** https://youtrack.raccoongang.com/issue/AXIMST-11

### CODE REVIEW

File: src/course-unit/header-navigations/HeaderNavigations.jsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  return (
+    <nav className="header-navigations ml-auto flex-shrink-0">
+      <Button
+        variant="outline-primary"
+        onClick={handleViewLive}
+      >
+        {intl.formatMessage(messages.viewLiveButton)}
+      </Button>
+      <Button
+        variant="outline-primary ml-2.5"
```
Comment (PKulkoRaccoonGang) [inline src/course-unit/header-navigations/HeaderNavigations.jsx:21]: [important] This margin adds unwanted space between buttons (correct 12 px). Let's remove it.

image

**Review by PKulkoRaccoonGang — COMMENTED** [review]
@ihor-romaniuk Looks great! An open question on the border-bottom of the `SubHeader` component. It is missing from the design for this page.

File: src/course-unit/CourseUnit.jsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/CourseUnit.jsx:1]: Agree with you, removed

File: src/CourseAuthoringRoutes.jsx [hunk 3]
```diff
@@ -16,6 +15,7 @@ import ScheduleAndDetails from './schedule-and-details';
 import { GradingSettings } from './grading-settings';
 import CourseTeam from './course-team/CourseTeam';
 import { CourseUpdates } from './course-updates';
+import CourseUnit from './course-unit/CourseUnit';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/CourseAuthoringRoutes.jsx:18]: Simplified

File: src/course-unit/CourseUnit.jsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import { getProcessingNotification } from '../generic/processing-notification/data/selectors';
+
+import HeaderTitle from './header-title/HeaderTitle';
+import Breadcrumbs from './breadcrumbs/Breadcrumbs';
+import HeaderNavigations from './header-navigations/HeaderNavigations';
+import SubHeader from '../generic/sub-header/SubHeader';
+
+import { useCourseUnit } from './hooks';
+import messages from './messages';
+import './CourseUnit.scss';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/CourseUnit.jsx:22]: Changed

File: src/course-unit/CourseUnit.test.jsx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    await waitFor(() => {
+      const viewLiveButton = getByRole('button', { name: headerNavigationsMessages.viewLiveButton.defaultMessage });
+      userEvent.click(viewLiveButton);
+      expect(window.open).toHaveBeenCalled();
+      expect(window.open).toHaveBeenCalledWith(`${getConfig().LMS_BASE_URL}/courses/${courseId}/jump_to/${blockId}`, '_blank');
+
+      const previewButton = getByRole('button', { name: headerNavigationsMessages.previewButton.defaultMessage });
+      userEvent.click(previewButton);
+      expect(window.open).toHaveBeenCalled();
+      expect(window.open).toHaveBeenCalledWith(`${getConfig().PREVIEW_BASE_URL}/courses/${courseId}/courseware/interactive_demonstrations/${sectionId}/1?activate_block_id=${blockId}`, '_blank');
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/CourseUnit.test.jsx:97]: Changed

File: src/course-unit/CourseUnit.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/CourseUnit.jsx:1]: Done

File: src/CourseAuthoringRoutes.jsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/CourseAuthoringRoutes.jsx:18]: Done

File: src/course-unit/breadcrumbs/Breadcrumbs.jsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import { useSelector } from 'react-redux';
+
+import { Dropdown, Icon } from '@edx/paragon';
+import {
+  ArrowDropDown as ArrowDropDownIcon,
+  ChevronRight as ChevronRightIcon,
+} from '@edx/paragon/icons';
+
+import { getCourseUnit } from '../data/selectors';
+import { useCourseOutline } from '../../course-outline/hooks';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/breadcrumbs/Breadcrumbs.jsx:12]: Done

File: src/generic/sub-header/SubHeader.scss [hunk 7]
```diff
@@ -1,5 +1,6 @@
 .sub-header {
   display: flex;
+  gap: 2rem;
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/generic/sub-header/SubHeader.scss:3]: Changed

File: src/course-unit/breadcrumbs/Breadcrumbs.jsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    subsection: {
+      id: ancestorInfo?.ancestors[0]?.id,
+      displayName: ancestorInfo?.ancestors[0]?.displayName,
+      dropdownItems: activeCourseSectionInfo?.childInfo.children || [],
+    },
+  };
+
+  return (
+    <div className="d-flex align-center mb-2.5">
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/breadcrumbs/Breadcrumbs.jsx:33]: We can use `nav` tag here but leave dropdown buttons as is. What do you think about it?

File: src/course-unit/breadcrumbs/Breadcrumbs.jsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        <Dropdown.Toggle id="breadcrumbs-dropdown-section" variant="link" className="p-0 text-primary small">
+          <span className="small">{breadcrumbs.section.displayName}</span>
+          <Icon
+            src={ArrowDropDownIcon}
+            className="text-primary ml-1"
+          />
+        </Dropdown.Toggle>
+        <Dropdown.Menu>
+          {breadcrumbs.section.dropdownItems.map(({ studioUrl, displayName }) => (
+            <Dropdown.Item
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/breadcrumbs/Breadcrumbs.jsx:44]: Removed `autoClose="inside"` to have auto-closing to click inside and outside as well.

File: src/course-unit/breadcrumbs/Breadcrumbs.jsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+      displayName: ancestorInfo?.ancestors[0]?.displayName,
+      dropdownItems: activeCourseSectionInfo?.childInfo.children || [],
+    },
+  };
+
+  return (
+    <div className="d-flex align-center mb-2.5">
+      <Dropdown autoClose="outside">
+        <Dropdown.Toggle id="breadcrumbs-dropdown-section" variant="link" className="p-0 text-primary small">
+          <span className="small">{breadcrumbs.section.displayName}</span>
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/breadcrumbs/Breadcrumbs.jsx:36]: Agree, thanks

File: src/course-unit/header-navigations/HeaderNavigations.jsx (hunk 1 above)
Comment (ihor-romaniuk) [reply src/course-unit/header-navigations/HeaderNavigations.jsx:21]: Fixed

File: src/course-unit/data/api.js [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import { camelCaseObject, getConfig } from '@edx/frontend-platform';
+import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+
+const getApiBaseUrl = () => getConfig().STUDIO_BASE_URL;
+
+/**
+ * Get course unit API url
+ * @param {string} itemId
+ * @returns {`${string}/xblock/container/${string}`}
+ */
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/api.js:11]: Done

File: src/course-unit/data/api.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+ * @returns {`${string}/xblock/${string}`}
+ */
+export const getXBlockBaseApiUrl = (itemId) => `${getApiBaseUrl()}/xblock/${itemId}`;
+
+/**
+ * Get course unit.
+ * @param {string} unitId
+ * @returns {Promise<Object>}
+ */
+export async function getCourseUnit(unitId) {
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/api.js:26]: Done

File: src/course-unit/data/api.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  const { data } = await getAuthenticatedHttpClient()
+    .get(getCourseUnitApiUrl(unitId));
+
+  return camelCaseObject(data);
+}
+
+/**
+ * Edit course unit.
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/api.js:34]: Done

File: src/course-unit/data/slice.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+/* eslint-disable no-param-reassign */
+import { createSlice } from '@reduxjs/toolkit';
+import { RequestStatus } from '../../data/constants';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/slice.js:3]: Done

File: src/course-unit/data/slice.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+export const {
+  fetchCourseItemSuccess,
+  updateLoadingCourseUnitStatus,
+  updateSavingStatus,
+} = slice.actions;
+
+export const {
+  reducer,
+} = slice;
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/slice.js:38]: Let's move it as is

File: src/course-unit/data/thunk.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  updateLoadingCourseUnitStatus,
+  fetchCourseItemSuccess,
+  updateSavingStatus,
+} from './slice';
+import { RequestStatus } from '../../data/constants';
+import { NOTIFICATION_MESSAGES } from '../../constants';
+import {
+  hideProcessingNotification,
+  showProcessingNotification,
+} from '../../generic/processing-notification/data/slice';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/thunk.js:12]: Done

File: src/course-unit/data/thunk.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/data/thunk.js:14]: Done

File: src/course-unit/header-navigations/HeaderNavigations.test.jsx [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+    expect(getByRole('button', { name: messages.viewLiveButton.defaultMessage })).toBeInTheDocument();
+    expect(getByRole('button', { name: messages.previewButton.defaultMessage })).toBeInTheDocument();
+  });
+
+  it('calls the correct handlers when clicking buttons', () => {
+    const { getByRole } = renderComponent();
+
+    const viewLiveButton = getByRole('button', { name: messages.viewLiveButton.defaultMessage });
+    fireEvent.click(viewLiveButton);
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/header-navigations/HeaderNavigations.test.jsx:36]: I cover this functionality with `userEvent` in the component unit test

File: src/course-unit/header-title/HeaderTitle.jsx [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+        className="ml-1 flex-shrink-0"
+        iconAs={EditIcon}
+        onClick={handleTitleEdit}
+      />
+      <IconButton
+        alt={intl.formatMessage(messages.altButtonSettings)}
+        className="flex-shrink-0"
+        iconAs={SettingsIcon}
+        onClick={() => {
+        }}
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/header-title/HeaderTitle.jsx:55]: Done

File: src/course-unit/header-title/messages.js [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import { defineMessages } from '@edx/frontend-platform/i18n';
+
+const messages = defineMessages({
+  altButtonEdit: {
+    id: 'course-authoring.course-unit.heading.button.edit.alt',
+    defaultMessage: 'Edit',
+  },
+  ariaLabelButtonEdit: {
+    id: 'course-authoring.course-unit.heading.button.Edit.ariaLabel',
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/header-title/messages.js:9]: Done

File: src/course-unit/header-title/messages.js [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+    defaultMessage: 'Edit field',
+  },
+  altButtonSettings: {
+    id: 'course-authoring.course-unit.heading.button.Settings.alt',
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/header-title/messages.js:13]: Done

File: src/course-unit/hooks.jsx [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+  fetchCourseUnitQuery,
+  editCourseItemQuery,
+} from './data/thunk';
+import {
+  getCourseUnit,
+  getLoadingStatus,
+  getSavingStatus,
+} from './data/selectors';
+import { updateSavingStatus } from './data/slice';
+import { RequestStatus } from '../data/constants';
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/hooks.jsx:15]: Done

File: src/course-unit/hooks.jsx [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+
+  const unitTitle = courseUnit.metadata?.displayName || '';
+
+  const headerNavigationsActions = {
+    handleViewLive: () => {
+      window.open(`${config.LMS_BASE_URL}/courses/${courseId}/jump_to/${blockId}`, '_blank');
+    },
+    handlePreview: () => {
+      const sectionId = courseUnit.ancestorInfo?.ancestors[0]?.id.split('@').pop();
+      window.open(`${config.PREVIEW_BASE_URL}/courses/${courseId}/courseware/interactive_demonstrations/${sectionId}/1?activate_block_id=${blockId}`, '_blank');
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/course-unit/hooks.jsx:35]: Done

File: src/generic/sub-header/SubHeader.jsx [hunk 24]
```diff
@@ -16,7 +16,7 @@ const SubHeader = ({
   <div className={`${!hideBorder && 'border-bottom border-light-400'} mb-3`}>
     <header className="sub-header">
       <h2 className="sub-header-title">
-        <small className="sub-header-title-subtitle">{subtitle}</small>
+        <div className="sub-header-title-subtitle">{subtitle}</div>
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/generic/sub-header/SubHeader.jsx:19]: Returned

File: src/generic/sub-header/SubHeader.jsx [hunk 25]
```diff
@@ -25,7 +25,7 @@ const SubHeader = ({
         )}
       </h2>
       {headerActions && (
-        <ActionRow className="ml-auto sub-header-actions">
+        <ActionRow className="ml-auto flex-shrink-0 sub-header-actions">
```
[the remark this answers is not in the document]
Comment (ihor-romaniuk) [reply src/generic/sub-header/SubHeader.jsx:28]: It should be added

**Review by ruzniaievdm — APPROVED** [review]

**Review by PKulkoRaccoonGang — APPROVED** [review]

File: src/course-unit/CourseUnit.jsx [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+  document.title = getPageHeadTitle('', unitTitle);
+
+  const {
+    isShow: isShowProcessingNotification,
+    title: processingNotificationTitle,
+  } = useSelector(getProcessingNotification);
+
+  if (isLoading) {
+    // eslint-disable-next-line react/jsx-no-useless-fragment
+    return <></>;
```
Comment (monteri) [inline src/course-unit/CourseUnit.jsx:46]: Maybe `return null`?

Comment (ihor-romaniuk) [reply src/course-unit/CourseUnit.jsx]: Done

File: src/course-unit/utils.jsx [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+);
+
+/**
+ * Method to return course unit preview URL path.
+ * @param {string} courseId
+ * @param {string} sectionId
+ * @param {string} blockId
+ * @returns {string} {`/courses/${courseId}/courseware/interactive_demonstrations/${sectionId}/1?activate_block_id=${blockId}`}
+ */
+export const getUnitPreviewPath = (courseId, sectionId, blockId) => (
```
Comment (GlugovGrGlib) [inline src/course-unit/utils.jsx:21]: @ihor-romaniuk 
I have checked, and we would need to take both links from backend
 https://github.com/raccoongang/edx-platform/pull/2484/files#diff-fcd65c5bab25478ab25e92d58b53a090ec0141ef462e24bd88052917dba52cafR73

Ticket: [AXIMST-221](https://youtrack.raccoongang.com/agiles/104-1224/105-3378?issue=AXIMST-221) Use view live and preview links from backend

**Pull request merged**